### ISSUE TITLE
Function Search Page API

### ISSUE THREAD

**Issue by subdavis** [microbiomedata/nmdc-server#191]
## Query params

As before, everything that accepts a query will follow the same schema.  Here is a declaration of all the schemas.

New bits are maked `NEW`, and the old bits are kept for contextual relevence.

``` typescript
interface Condition {
  field: string;
  op: opType;
  value: string | number | [number, number];
  table: string;
}

interface SearchParams {
  offset?: number;
  limit?: number;
  conditions: Condition[];
}

interface SearchResponse<T> {
  count: number;
  results: T[];
}

interface BaseSearchResult {
  id: string;
  name: string;
  description: string;
  alternate_ideantifiers: string[];
  annotations: Record<string, unknown>;
}

interface StudySearchResults extends BaseSearchResult {
  principal_investigator_websites: string[];
  publication_dois: string[];
  gold_name: string;
  gold_description: string;
  scientific_objective: string;
  add_date: string;
  mod_date: string;
  open_in_gold: string;
  /* NEW */
  principal_investigator_name: string;
  /* NEW  possible aggregations.  Maybe they update with query params, maybe they don't. */
  sample_count: number;
  omics_counts: {
    type: 'proteomics' | 'metabolomics' | 'whateveromics';
    count: number;
  }[];
}

/* NEW */
interface OmicsDataObject {
  filename: string;
  description: string;
  size_bytes: number;
  download_uri: string; // or url, or whatever is most convenient.
  data_object_type: 'raw' | 'qc' | 'whatever';
}

/* NEW */
interface OmicsPipeline {
  name: string;
  data_objects: OmicsDataObject[];
}

/* NEW */
interface SampleOmicsData {
  type: 'proteomics' | 'metabolomics' | 'whateveromics';
  // The number of runs total
  total_count: number;
  // The number of runs matching the search criteria.
  // I would think that only KO term actually causes this to be different than total_count,
  // but I can imagine that "Omics Processing Type" facets will be important on this page.
  // So if I filter samples by "Metabolomics", then all the other omics data on each sample should be
  // visible but grayed out.   /shrug?
  query_matches_count: number;
  // TODO: name this something better
  omics_pipipelines: OmicsPipeline[];
}

interface BiosampleSearchResult extends BaseSearchResult {
  project_id: string;
  depth: number;
  env_broad_scale_id: string;
  env_local_scale_id: string;
  env_medium_id: string;
  longitude: number;
  latitude: number;
  add_date: string;
  mod_date: string;
  open_in_gold: string;
  env_broad_scale: {
    id: string;
    label: string;
    data: string;
  };
  env_local_scale: {
    id: string;
    label: string;
    data: string;
  };
  env_medium: {
    id: string;
    label: string;
    data: string;
  };
  /* NEW.  I'd prefer to just have all of this in one go,
     but if this needs to be a URI list such that additional requests return
     SampleOmicsData objects, I can deal with that. 
  */
  omics_data: SampleOmicsData[];
}
```

## Study API

I don't think the study list is filtered by the active query.  I believed that there would be aggregations of omics data at the study level, but I don't see that in our discussion logs.

`GET /<prefix>/study` accepts `SearchParams`

Returns `SearchResponse<StudySearchResult>`

## Sample API

`GET /<prefix>/sample` accepts `SearchParams`

Returns `SearchResponse<SampleSearchResult>`

## Visualizations and Aggregations

I'll follow up with this soon.  I don't love having to make 2-3 queries for every vis element (1 without search filters, one with all except the facet to which the vis belongs, and one with all filters).  It'd be nice to get all of that  in 1 request, but it would either be more expensive or push the burden of caching into the server.

I maybe need to implement caching more intelligently on the client.  Either way, I'll come back and update this soon.

**Comment by jeffbaumes** [comment]
> I believed that there would be aggregations of omics data at the study level, but I don't see that in our discussion logs.

If it is an easy lift we may want to have some simple aggregations (# of samples that match the search, # of each omics type), but also ok for now to just have the study name and PI displayed.

**Comment by subdavis** [comment]
Updated `StudySearchResults` to match.

**Comment by kfagnan** [comment]
I think the aggregations of omics data at the study level came up in prior
discussions and will be valuable to the users. Agree that it can be added
later if it's too much right now.

**Comment by kfagnan** [comment]
Retraction - we will need the simple aggregations of omics data at the
study level to provide the statistics to help users navigate the data
portal - as discussed in the hierarchy issue.

**Comment by jbeezley** [comment]
Just a few minor corrections now that I'm working on implementing this.
* `principal_investigator_name` is not a new attribute
* I'm 99% sure what you mean by `OmicsPipeline` is actually [`OmicsProcessing`](https://microbiomedata.github.io/nmdc-metadata/OmicsProcessing/) which in the api is `Project`.  (We should probably prioritize changing that once this sprint is done.)
* `OmicsDataObject` will probably just be [`DataObject`](https://microbiomedata.github.io/nmdc-metadata/DataObject/).  I'm imagining endpoints such as `/data-object?workflow-execution=<id>` and `/data-object/:id/download` (for the redirect that store statistics) as being useful generally.
* `SampleOmicsData` is most closely related to a [`WorkflowExecutionActivity`](https://microbiomedata.github.io/nmdc-metadata/WorkflowExecutionActivity/).  As such, I would probably name `omics_pipelines` as `workflow_executions` or `workflow_execution_activities`.

For `sample_count` and `omics_count`, what information do we need to decide if the current query should reflect this count?  The implementations will most likely be significantly different.  If it's a constant value, I think I'll denormalize it in the database.  If not, it will have to get baked into the query in a relatively complicated way.

**Comment by subdavis** [comment]
> * `principal_investigator_name` is not a new attribute

cool

> * I'm 99% sure what you mean by `OmicsPipeline` is actually [`OmicsProcessing`]

Looks right to me

> * `OmicsDataObject` will probably just be [`DataObject`]

As long as it includes some version of those fields, that's fine.

> I'm imagining endpoints such as `/data-object?workflow-execution=<id>`

It would be nice to not have the extra layer of api calls.

I was hoping for...

* 1 call to get the list of samples (including aggregations about their omics processing summaries)
* 1 additional call for each sample, which gives a list of all omics processing pipelines including all files for each.

So something more like `/omics_processing?sample_id=<id>` which includes all data objects for all ~~SampleOmicsData~~ Workflow Execution Activities.

See below.

I think what you're describing would require many calls to see the results for a single sample.

> `SampleOmicsData` is most closely related to a [`WorkflowExecutionActivity`](

Now I'm confused.  `WorkflowExecutionActivity` has no relationship with `OmicsProcessing` in the docs, but in my models, `SampleOmicsData` is a list of `OmicsPipeline` sharing a common omics type.   

How do these things relate?

**Comment by jeffbaumes** [comment]
> For sample_count and omics_count, what information do we need to decide if the current query should reflect this count? The implementations will most likely be significantly different. If it's a constant value, I think I'll denormalize it in the database. If not, it will have to get baked into the query in a relatively complicated way.

@kfagnan mentioned the below on #190, so I believe the answer is that it is important to summarize the sample and omics types counts **taking into account the current search**, so denormalization would not be an option here:

> There is also consensus on the need for a the aggregated omics information
> under studies. The ask if if you can take the tiles from the home page -
> #studies, #metagenomes, #metaproteomes, etc and carry that through to each
> landing page.
>
> Some users will want to know if there are proteomics data sets available
> for a given KO search term and we want this summary stats information to be
> prominent.

**Comment by jbeezley** [comment]
This was implemented in #197 and #198
